File: Scripts/update-lists.py
#!/usr/bin/env python3
"""
Automated Blocklist Updater for Ven0m0-Adblock
Downloads and updates filter lists from remote URLs, validates checksums,
and maintains source tracking metadata.
"""

import argparse
import asyncio
import base64
import hashlib
import io
import json
import logging
import re
import sys
import tempfile
from pathlib import Path
from typing import Final
import aiohttp
import aiofiles

# Import common utilities
if str(Path(__file__).parent) not in sys.path:
    sys.path.append(str(Path(__file__).parent))

from common import sanitize_filename

# ============================================================================
# CONFIGURATION
# ============================================================================
SOURCES_CONFIG: Final[str] = "lists/sources-urls.json"
DEFAULT_OUTPUT: Final[str] = "lists/sources"
METADATA_FILE: Final[str] = "lists/sources-metadata.json"
TIMEOUT: Final[int] = 60
CHUNK_SIZE: Final[int] = 65536
MAX_CONCURRENT: Final[int] = 10

# ============================================================================
# LOGGING
# ============================================================================
logging.basicConfig(
  level=logging.INFO,
  format="[%(asctime)s] %(levelname)s %(message)s",
  datefmt="%H:%M:%S",
)
logger = logging.getLogger(__name__)

# ============================================================================
# CHECKSUM VALIDATION
# ============================================================================

def validate_checksum(content: str, name: str = "unknown") -> bool:
  """Validate Adblock Plus checksum header."""
  pattern = re.compile(
    r"^\s*!\s*checksum[\s\-:]+([\w\+\/=]+).*\n",
    re.MULTILINE | re.IGNORECASE,
  )
  match = pattern.search(content)

  if not match:
    logger.debug(f"No checksum in {name} (optional)")
    return True

  declared_checksum = match.group(1)
  data_no_checksum = pattern.sub("", content, 1)
  normalized = data_no_checksum.replace("\r", "").rstrip("\n") + "\n"
  computed_hash = hashlib.sha256(normalized.encode("utf-8")).digest()
  computed_checksum = base64.b64encode(computed_hash).decode().rstrip("=")

  if declared_checksum == computed_checksum:
    logger.info(f"✓ Checksum valid: {name}")
    return True

  logger.error(
    f"✗ Checksum mismatch in {name}: "
    f"expected {computed_checksum}, got {declared_checksum}"
  )
  return False

# ============================================================================
# FILE OPERATIONS
# ============================================================================

def count_rules(content: str) -> int:
  """Count active rules in the content."""
  return sum(
    1
    for line in io.StringIO(content)
    if (stripped := line.strip())
    and not stripped.startswith(("! ", "#", "["))
  )

async def process_downloaded_file(
  temp_path: Path,
  url: str,
  filename: str,
  output_dir: Path,
  skip_checksum: bool = False,
) -> Path | None:
  """Process and move temp file to final destination."""
  dest_path = output_dir / filename

  try:
    async with aiofiles.open(temp_path, mode="r", encoding="utf-8") as f:
      content = await f.read()

    if not skip_checksum:
      # Offload CPU-bound checksum validation to a thread
      is_valid = await asyncio.to_thread(validate_checksum, content, filename)
      if not is_valid:
        logger.warning(f"Checksum validation failed for {url}")
        # Do not delete temp_path here; the caller's finally block is responsible for cleanup.
        return None
    
    if len(content) < 100:
      logger.error(f"Downloaded file suspiciously small ({len(content)} bytes): {url}")
      return None
    
    # Offload CPU-bound rule counting to a thread
    rule_count = await asyncio.to_thread(count_rules, content)
    
    async with aiofiles.open(dest_path, mode="w", encoding="utf-8") as f:
      await f.write(content)

    logger.info(f"✓ {filename} ({rule_count} rules)")
    return dest_path

  except Exception as e:
    logger.exception(f"Error processing {url}: {e}")
    if temp_path.exists():
      temp_path.unlink()
    return None

# ============================================================================
# ASYNC DOWNLOAD
# ============================================================================

async def fetch_list(
  session: aiohttp.ClientSession,
  url: str,
  filename: str,
  output_dir: Path,
  skip_checksum: bool = False,
) -> tuple[str, bool]:
  """Download a single filter list."""
  try:
    headers = {
      "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36",
      "Accept": "text/plain,*/*",
    }
    
    async with session.get(url, timeout=TIMEOUT, headers=headers) as resp:
      resp.raise_for_status()

      tmp_path = None
      with tempfile.NamedTemporaryFile(
        mode="wb",
        delete=False,
        suffix=".txt",
      ) as tmp:
        tmp_path = Path(tmp.name)
        # Close immediately, re-open async below

      try:
        async with aiofiles.open(tmp_path, mode="wb") as f:
          async for chunk in resp.content.iter_chunked(CHUNK_SIZE):
            await f.write(chunk)

        # Only call process once with the correct filename
        result = await process_downloaded_file(tmp_path, url, filename, output_dir, skip_checksum)
        return (url, result is not None)
      finally:
        # Ensure cleanup always
        if tmp_path:
          try:
            await asyncio.to_thread(tmp_path.unlink)
          except FileNotFoundError:
            pass

  except asyncio.TimeoutError:
    logger.error(f"✗ Timeout: {url}")
  except aiohttp.ClientError as e:
    logger.error(f"✗ HTTP error for {url}: {e}")
  except Exception as e:
    logger.exception(f"✗ Unexpected error for {url}: {e}")

  return (url, False)

# ============================================================================
# SOURCE CONFIGURATION
# ============================================================================

def load_sources(config_path: Path) -> dict[str, dict]:
  """Load source URLs configuration."""
  if not config_path.exists():
    logger.warning(f"Config not found: {config_path}, creating template")
    template = {
      "sources": [
        {
          "url": "https://raw.githubusercontent.com/uBlockOrigin/uAssets/master/filters/filters.txt",
          "filename": "uBlock-Filters.txt",
          "skip_checksum": False,
          "enabled": True
        },
        {
          "url": "https://easylist.to/easylist/easylist.txt",
          "filename": "EasyList.txt",
          "skip_checksum": False,
          "enabled": True
        }
      ]
    }
    config_path.parent.mkdir(parents=True, exist_ok=True)
    config_path.write_text(json.dumps(template, indent=2) + "\n", encoding="utf-8")
    logger.info(f"Created template config: {config_path}")
  
  data = json.loads(config_path.read_text(encoding="utf-8"))
  return {
    src["url"]: {
      "filename": src.get("filename") or sanitize_filename(src["url"]),
      "skip_checksum": src.get("skip_checksum", False),
      "enabled": src.get("enabled", True),
    }
    for src in data.get("sources", [])
    if src.get("enabled", True)
  }

def save_metadata(sources: dict, results: dict[str, bool], output_dir: Path) -> None:
  """Save download metadata for tracking."""
  from datetime import datetime, timezone
  
  metadata = {
    "last_updated": datetime.now(timezone.utc).isoformat(),
    "sources": {
      url: {
        "filename": config["filename"],
        "success": results.get(url, False),
        "checksum_validated": not config["skip_checksum"],
      }
      for url, config in sources.items()
    }
  }

  metadata_path = Path(METADATA_FILE)
  metadata_path.write_text(
    json.dumps(metadata, indent=2, sort_keys=True) + "\n",
    encoding="utf-8",
  )
  logger.info(f"Saved metadata: {metadata_path}")

# ============================================================================
# MAIN PIPELINE
# ============================================================================

async def main() -> int:
  """Main execution flow."""
  parser = argparse.ArgumentParser(
    description="Update Ven0m0-Adblock filter lists from remote sources"
  )
  parser.add_argument(
    "--config",
    type=Path,
    default=SOURCES_CONFIG,
    help="Source URLs configuration file (JSON)",
  )
  parser.add_argument(
    "--output-dir",
    type=Path,
    default=DEFAULT_OUTPUT,
    help="Output directory for downloaded lists",
  )
  parser.add_argument(
    "--max-concurrent",
    type=int,
    default=MAX_CONCURRENT,
    help="Max concurrent downloads",
  )
  parser.add_argument(
    "--filter",
    help="Only update sources matching this substring",
  )
  parser.add_argument(
    "--validate",
    action="store_true",
    help="Run AGLint validation after downloads",
  )
  args = parser.parse_args()

  output_dir: Path = args.output_dir
  output_dir.mkdir(parents=True, exist_ok=True)

  logger.info("Loading source configuration...")
  sources = load_sources(args.config)
  
  if args.filter:
    sources = {
      url: cfg for url, cfg in sources.items()
      if args.filter.lower() in url.lower() or args.filter.lower() in cfg["filename"].lower()
    }
    logger.info(f"Filtered to {len(sources)} sources matching '{args.filter}'")

  if not sources:
    logger.error("No sources to update")
    return 1

  logger.info(f"Updating {len(sources)} filter lists...")
  
  connector = aiohttp.TCPConnector(limit=args.max_concurrent)
  async with aiohttp.ClientSession(connector=connector) as session:
    tasks = [
      fetch_list(session, url, cfg["filename"], output_dir, cfg["skip_checksum"])
      for url, cfg in sources.items()
    ]
    results = await asyncio.gather(*tasks, return_exceptions=False)

  results_dict = dict(results)
  success_count = sum(1 for success in results_dict.values() if success)
  
  save_metadata(sources, results_dict, output_dir)
  
  logger.info(f"✓ Updated {success_count}/{len(sources)} lists successfully")

  if args.validate:
    try:
      import subprocess
      logger.info("Running AGLint validation...")
      result = subprocess.run(
        ["bun", "x", "aglint", str(output_dir / "*.txt")],
        capture_output=True,
        text=True,
        check=False,
      )
      if result.returncode != 0:
        logger.warning("AGLint found issues (non-blocking)")
    except Exception as e:
      logger.warning(f"Could not run AGLint: {e}")

  return 0 if success_count == len(sources) else 1

if __name__ == "__main__":
  sys.exit(asyncio.run(main()))

File: Scripts/update_lists.ts
#!/usr/bin/env bun
/**
 * Automated Blocklist Updater for Ven0m0-Adblock
 * Downloads and updates filter lists from remote URLs, validates checksums,
 * and maintains source tracking metadata.
 */

import { parseArgs } from "util";
import { createHash } from "node:crypto";
import { mkdir, rm, rename, mkdtemp, unlink } from "node:fs/promises";
import { join, dirname, basename } from "node:path";
import { tmpdir } from "node:os";
import { sanitizeFilename } from "./common.ts";

// ============================================================================
// CONFIGURATION
// ============================================================================
const SOURCES_CONFIG = "lists/sources-urls.json";
const DEFAULT_OUTPUT = "lists/sources";
const METADATA_FILE = "lists/sources-metadata.json";
const HEADER_PREFIXES = ["! ", "#", "["];
const TIMEOUT_MS = 60000;
const MAX_CONCURRENT = 10;

// ============================================================================
// LOGGING
// ============================================================================
const logger = {
    info: (msg: string) => console.log(`[${new Date().toLocaleTimeString('en-US', { hour12: false })}] INFO ${msg}`),
    warning: (msg: string) => console.warn(`[${new Date().toLocaleTimeString('en-US', { hour12: false })}] WARNING ${msg}`),
    error: (msg: string) => console.error(`[${new Date().toLocaleTimeString('en-US', { hour12: false })}] ERROR ${msg}`),
};

// ============================================================================
// CHECKSUM VALIDATION
// ============================================================================

export function validateChecksum(content: string, name: string): boolean {
    const lines = content.split(/\r?\n/);
    let declaredChecksum = "";
    let dataStart = 0;

    for (let i = 0; i < Math.min(lines.length, 15); i++) {
        const line = lines[i].trim();
        const match = line.match(/^!\s*Checksum:\s*([\w+\/=]+)/i);
        if (match) {
            declaredChecksum = match[1];
            dataStart = i + 1;
            break;
        }
    }

    if (!declaredChecksum) {
        logger.warning(`No checksum found in ${name}`);
        return true; // No checksum to validate
    }

    const dataToHash = lines.slice(dataStart).join("\n") + (lines.length > dataStart ? "\n" : "");

    // Normalize newlines and encode to utf-8 before hashing
    const normalized = dataToHash.replace(/\r\n/g, "\n");
    const computedHash = createHash("md5").update(normalized, "utf-8").digest();

    // Encode to base64 and strip padding to match typical Adblock formatting
    const computedChecksum = computedHash.toString("base64").replace(/=+$/, "");

    if (declaredChecksum === computedChecksum) {
        logger.info(`✓ Checksum valid: ${name}`);
        return true;
    }

    logger.error(`✗ Checksum mismatch in ${name}: expected ${computedChecksum}, got ${declaredChecksum}`);
    return false;
}

// ============================================================================
// FILE OPERATIONS
// ============================================================================

export function countRules(content: string): number {
    let count = 0;
    const lines = content.split(/\r?\n/);
    for (const line of lines) {
        const stripped = line.trim();
        if (stripped && !HEADER_PREFIXES.some(p => stripped.startsWith(p))) {
            count++;
        }
    }
    return count;
}

export async function processDownloadedFile(
    tempPath: string,
    url: string,
    filename: string,
    outputDir: string,
    skipChecksum = false
): Promise<string | null> {
    const destPath = join(outputDir, basename(filename));

    try {
        const file = Bun.file(tempPath);
        const content = await file.text();

        if (!skipChecksum) {
            const isValid = validateChecksum(content, filename);
            if (!isValid) {
                logger.warning(`Checksum validation failed for ${url}`);
                return null;
            }
        }

        if (content.length < 100) {
            logger.error(`Downloaded file suspiciously small (${content.length} bytes): ${url}`);
            return null;
        }

        const ruleCount = countRules(content);

        await Bun.write(destPath, content);
        logger.info(`✓ ${filename} (${ruleCount} rules)`);
        return destPath;
    } catch (e: any) {
        logger.error(`Error processing ${url}: ${e.message || e}`);
        return null;
    }
}

// ============================================================================
// ASYNC DOWNLOAD
// ============================================================================

export async function fetchList(
    url: string,
    filename: string,
    outputDir: string,
    skipChecksum = false
): Promise<[string, boolean]> {
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), TIMEOUT_MS);

        const headers = new Headers({
            "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36",
            "Accept": "text/plain,*/*",
        });

        const resp = await fetch(url, { headers, signal: controller.signal });
        clearTimeout(timeoutId);

        if (!resp.ok) {
            throw new Error(`HTTP ${resp.status} ${resp.statusText}`);
        }

        const tempDirPath = await mkdtemp(join(tmpdir(), "al-"));
        const tmpPath = join(tempDirPath, "tmp.txt");

        try {
            await Bun.write(tmpPath, resp);

            const result = await processDownloadedFile(tmpPath, url, filename, outputDir, skipChecksum);
            return [url, result !== null];
        } finally {
            await unlink(tmpPath).catch(() => {});
            await rm(tempDirPath, { recursive: true, force: true }).catch(() => {});
        }
    } catch (e: any) {
        if (e.name === 'AbortError') {
            logger.error(`✗ Timeout: ${url}`);
        } else {
            logger.error(`✗ Error for ${url}: ${e.message || e}`);
        }
    }

    return [url, false];
}

// ============================================================================
// SOURCE CONFIGURATION
// ============================================================================

export async function loadSources(configPath: string): Promise<Record<string, any>> {
    const file = Bun.file(configPath);
    if (!(await file.exists())) {
        logger.warning(`Config not found: ${configPath}, creating template`);
        const template = {
            sources: [
                {
                    url: "https://raw.githubusercontent.com/uBlockOrigin/uAssets/master/filters/filters.txt",
                    filename: "uBlock-Filters.txt",
                    skip_checksum: false,
                    enabled: true,
                },
                {
                    url: "https://easylist.to/easylist/easylist.txt",
                    filename: "EasyList.txt",
                    skip_checksum: false,
                    enabled: true,
                },
            ]
        };
        await mkdir(dirname(configPath), { recursive: true });
        await Bun.write(configPath, JSON.stringify(template, null, 2) + "\n");
        logger.info(`Created template config: ${configPath}`);
    }

    const content = await Bun.file(configPath).text();
    const data = JSON.parse(content);
    const result: Record<string, any> = {};

    for (const src of (data.sources || [])) {
        if (src.enabled !== false) {
            result[src.url] = {
                filename: src.filename || sanitizeFilename(src.url),
                skip_checksum: src.skip_checksum || false,
                enabled: true,
            };
        }
    }
    return result;
}

export async function saveMetadata(
    sources: Record<string, any>,
    results: Record<string, boolean>,
    outputDir: string
): Promise<void> {
    const metadata = {
        last_updated: new Date().toISOString(),
        sources: {} as Record<string, any>,
    };

    for (const [url, config] of Object.entries(sources)) {
        metadata.sources[url] = {
            filename: config.filename,
            success: results[url] || false,
            checksum_validated: !config.skip_checksum,
        };
    }

    const metadataPath = METADATA_FILE;
    await Bun.write(metadataPath, JSON.stringify(metadata, null, 2) + "\n");
    logger.info(`Saved metadata: ${metadataPath}`);
}

// ============================================================================
// MAIN PIPELINE
// ============================================================================

export async function main(): Promise<number> {
    const { values } = parseArgs({
        args: Bun.argv,
        options: {
            config: {
                type: "string",
                default: SOURCES_CONFIG,
            },
            "output-dir": {
                type: "string",
                default: DEFAULT_OUTPUT,
            },
            "max-concurrent": {
                type: "string",
                default: String(MAX_CONCURRENT),
            },
            filter: {
                type: "string",
            },
            validate: {
                type: "boolean",
            },
        },
        strict: true,
        allowPositionals: true,
    });

    const outputDir = values["output-dir"] as string;
    const maxConcurrent = parseInt(values["max-concurrent"] as string, 10);
    const filter = values.filter as string | undefined;

    await mkdir(outputDir, { recursive: true });

    logger.info("Loading source configuration...");
    let sources = await loadSources(values.config as string);

    if (filter) {
        const filteredSources: Record<string, any> = {};
        for (const [url, cfg] of Object.entries(sources)) {
            if (url.toLowerCase().includes(filter.toLowerCase()) ||
                cfg.filename.toLowerCase().includes(filter.toLowerCase())) {
                filteredSources[url] = cfg;
            }
        }
        sources = filteredSources;
        logger.info(`Filtered to ${Object.keys(sources).length} sources matching '${filter}'`);
    }

    const sourceKeys = Object.keys(sources);
    if (sourceKeys.length === 0) {
        logger.error("No sources to update");
        return 1;
    }

    logger.info(`Updating ${sourceKeys.length} filter lists...`);

    const resultsDict: Record<string, boolean> = {};

    // Batch processing
    for (let i = 0; i < sourceKeys.length; i += maxConcurrent) {
        const batch = sourceKeys.slice(i, i + maxConcurrent);
        const tasks = batch.map(url => {
            const cfg = sources[url];
            return fetchList(url, cfg.filename, outputDir, cfg.skip_checksum);
        });

        const results = await Promise.all(tasks);
        for (const [url, success] of results) {
            resultsDict[url] = success;
        }
    }

    const successCount = Object.values(resultsDict).filter(Boolean).length;
    await saveMetadata(sources, resultsDict, outputDir);

    logger.info(`✓ Updated ${successCount}/${sourceKeys.length} lists successfully`);

    if (values.validate) {
        try {
            logger.info("Running AGLint validation...");
            // Run bun x aglint path/to/files/*.txt
            const proc = Bun.spawn(["bun", "x", "aglint", `${outputDir}/*.txt`], {
                stdout: "pipe",
                stderr: "pipe",
            });
            const exitCode = await proc.exited;
            if (exitCode !== 0) {
                logger.warning("AGLint found issues (non-blocking)");
            }
        } catch (e: any) {
            logger.warning(`Could not run AGLint: ${e.message || e}`);
        }
    }

    return successCount === sourceKeys.length ? 0 : 1;
}

if (import.meta.main) {
    process.exitCode = await main();
}
